BUG: --lf running only a subset of last-failed tests

After running the pandas test suite and getting e.g. 20 test failures, running the same command with " --lf" appended should run those 20 tests.  In practice I'm seeing only fraction of these run on the subsequent run.

```
$ pytest pandas/tests --skip-slow --skip-db
[...]
=================================================================== short test summary info ====================================================================
FAILED pandas/tests/resample/test_resampler_grouper.py::test_apply_columns_multilevel - KeyError: Timestamp('2017-01-01 00:00:00', freq='H')
FAILED pandas/tests/groupby/test_function.py::test_median_empty_bins[True] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
FAILED pandas/tests/groupby/test_function.py::test_ops_general[count-size] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column name="0") are different
FAILED pandas/tests/groupby/test_grouping.py::TestGrouping::test_evaluate_with_empty_groups[agg-expected1] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>0] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>1] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>2] - AssertionError: Attributes of DataFrame.iloc[:, 0] (column ...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-mean-mean] - AssertionError: Attributes of DataFrame.iloc[:, 0] (col...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-median-<lambda>] - AssertionError: Attributes of DataFrame.iloc[:, 0...
FAILED pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-var-<lambda>] - AssertionError: Attributes of DataFrame.iloc[:, 0] (...
FAILED pandas/tests/groupby/aggregate/test_other.py::test_agg_lambda_with_timezone - ValueError: Must produce aggregated value
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[NotImplementedError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[RuntimeError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[KeyError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[IndexError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[OSError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ValueError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ArithmeticError] - AssertionError: Attributes of Series are different
FAILED pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[AttributeError] - AssertionError: Attributes of Series are different
====================================== 20 failed, 5034 passed, 1652 skipped, 374 xfailed, 1 warning in 208.51s (0:03:28) =======================================

$ pytest pandas/tests/resample pandas/tests/groupby pandas/tests/generic/ --skip-slow --lf
[...]
============================================================== 1 failed, 1455 deselected in 0.68s ==============================================================
```

Running pytest 6.0.2, py3.8.5, OSX 10.15.5

```
$ cat pytest_cache-bak/v/cache/lastfailed
{
  "pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg": true,
  "pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg_ea_method": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_aggregate_item_by_item": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>0]": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>1]": true,
  "pandas/tests/groupby/aggregate/test_aggregate.py::test_multiindex_custom_func[<lambda>2]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-mean-mean]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-median-<lambda>]": true,
  "pandas/tests/groupby/aggregate/test_cython.py::test_cython_agg_empty_buckets[True-var-<lambda>]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_agg_lambda_with_timezone": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ArithmeticError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[AttributeError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[IndexError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[KeyError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[NotImplementedError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[OSError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[RuntimeError]": true,
  "pandas/tests/groupby/aggregate/test_other.py::test_groupby_agg_err_catching[ValueError]": true,
  "pandas/tests/groupby/test_function.py::test_median_empty_bins[True]": true,
  "pandas/tests/groupby/test_function.py::test_ops_general[count-size]": true,
  "pandas/tests/groupby/test_grouping.py::TestGrouping::test_evaluate_with_empty_groups[agg-expected1]": true,
  "pandas/tests/resample/test_resampler_grouper.py::test_apply_columns_multilevel": true
}%
```

The output from the second run:

```
========================================================================= short test summary info ==========================================================================
FAILED pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg - AssertionError: Attributes of Series are different
FAILED pandas/tests/extension/decimal/test_decimal.py::test_groupby_agg_ea_method - AssertionError: Attributes of Series are different
============================================================= 2 failed, 30232 deselected, 2 warnings in 9.07s ==============================================================

```

--- a/changelog/7758.bugfix.rst
+++ b/changelog/7758.bugfix.rst
@@ -1,0 +1,1 @@
+Fixed an issue where some files in packages are getting lost from ``--lf`` even though they contain tests that failed. Regressed in pytest 5.4.0.

--- a/src/_pytest/cacheprovider.py
+++ b/src/_pytest/cacheprovider.py
@@ -28,6 +28,7 @@ from _pytest.config.argparsing import Parser
 from _pytest.fixtures import FixtureRequest
 from _pytest.main import Session
 from _pytest.python import Module
+from _pytest.python import Package
 from _pytest.reports import TestReport
 
 
@@ -232,7 +233,10 @@ class LFPluginCollSkipfiles:
     def pytest_make_collect_report(
         self, collector: nodes.Collector
     ) -> Optional[CollectReport]:
-        if isinstance(collector, Module):
+        # Packages are Modules, but _last_failed_paths only contains
+        # test-bearing paths and doesn't try to include the paths of their
+        # packages, so don't filter them.
+        if isinstance(collector, Module) and not isinstance(collector, Package):
             if Path(str(collector.fspath)) not in self.lfplugin._last_failed_paths:
                 self.lfplugin._skipped_files += 1
 

--- a/testing/test_cacheprovider.py
+++ b/testing/test_cacheprovider.py
@@ -7,6 +7,7 @@ import py
 
 import pytest
 from _pytest.config import ExitCode
+from _pytest.pytester import Pytester
 from _pytest.pytester import Testdir
 
 pytest_plugins = ("pytester",)
@@ -981,6 +982,36 @@ class TestLastFailed:
             consecutive=True,
         )
         assert result.ret == 0
+
+    def test_packages(self, pytester: Pytester) -> None:
+        """Regression test for #7758.
+
+        The particular issue here was that Package nodes were included in the
+        filtering, being themselves Modules for the __init__.py, even if they
+        had failed Modules in them.
+
+        The tests includes a test in an __init__.py file just to make sure the
+        fix doesn't somehow regress that, it is not critical for the issue.
+        """
+        pytester.makepyfile(
+            **{
+                "__init__.py": "",
+                "a/__init__.py": "def test_a_init(): assert False",
+                "a/test_one.py": "def test_1(): assert False",
+                "b/__init__.py": "",
+                "b/test_two.py": "def test_2(): assert False",
+            },
+        )
+        pytester.makeini(
+            """
+            [pytest]
+            python_files = *.py
+            """
+        )
+        result = pytester.runpytest()
+        result.assert_outcomes(failed=3)
+        result = pytester.runpytest("--lf")
+        result.assert_outcomes(failed=3)
 
 
 class TestNewFirst:

